### PR TITLE
Add cumulative chart for signed extrinsic

### PR DESCRIPTION
**Pull request by karthikeyan-cere**

### CODE REVIEW

File: backend/docker/sql/views/signed_extrinsics_per_day.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+CREATE VIEW signed_extrinsics_per_day AS
```
Comment (AndreiNavoichyk) [inline backend/docker/sql/views/signed_extrinsics_per_day.sql]: Add `view` postfix. `signed_extrinsics_per_day_view`

File: backend/docker/sql/views/signed_extrinsics_per_day.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+SELECT *
+FROM (SELECT "when"::date FROM generate_series(timestamp '2021-10-16', timestamp '2022-11-17', interval  '1 day') "when") d
```
Comment (AndreiNavoichyk) [inline backend/docker/sql/views/signed_extrinsics_per_day.sql]: No hardcode, please.

File: backend/docker/sql/views/signed_extrinsics_per_day.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+LEFT JOIN (
+  SELECT to_timestamp(timestamp)::date AS "when", count(*) AS "volume"
+  FROM extrinsic
+  WHERE is_signed=true
+  GROUP BY to_timestamp(timestamp)::date
```
Comment (AndreiNavoichyk) [inline backend/docker/sql/views/signed_extrinsics_per_day.sql:8]: Let's declare `to_timestamp(timestamp)` once.

File: backend/docker/sql/views/signed_extrinsics_per_month.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+CREATE view signed_extrinsics_per_month AS
```
Comment (AndreiNavoichyk) [inline backend/docker/sql/views/signed_extrinsics_per_month.sql]: Add `view` postfix. `signed_extrinsics_per_month_view`

File: frontend/locales/en.json [hunk 5]
```diff
@@ -66,6 +66,7 @@
       "nominators": "Nominators",
       "accounts": "Accounts",
       "faucet": "Faucet",
+      "cere-chart": "Chart",
```
Comment (AndreiNavoichyk) [inline frontend/locales/en.json]: Why do we need `cere` prefix here? We are going to have few charts. Rename to plural.

File: frontend/locales/en.json [hunk 6]
```diff
@@ -287,6 +288,11 @@
       "title": "Faucet",
       "head_title": "Cere Stats - Faucet",
       "head_content": "{networkName} Faucet"
+    },
+    "chart": {
```
Comment (AndreiNavoichyk) [inline frontend/locales/en.json]: Rename `chart` to plural.

File: frontend/locales/en.json [hunk 7]
```diff
@@ -287,6 +288,11 @@
...
+      "title": "Charts",
+      "head_title": "Cere Stats - Chart",
```
Comment (AndreiNavoichyk) [inline frontend/locales/en.json]: Rename `Chart` to plural.

File: frontend/locales/en.json [hunk 8]
```diff
@@ -287,6 +288,11 @@
...
+      "head_content": "Cere Stats - Chart"
```
Comment (AndreiNavoichyk) [inline frontend/locales/en.json]: Rename `Chart` to plural.

File: frontend/locales/es.json [hunk 9]
```diff
@@ -66,6 +66,7 @@
       "nominators": "Nominadores",
       "accounts": "Cuentas",
       "faucet": "Faucet",
+      "cere-chart": "Gráfico",
```
Comment (AndreiNavoichyk) [inline frontend/locales/es.json]: Adjust this file according to the en.json one.

File: frontend/package.json [hunk 10]
```diff
@@ -39,7 +39,7 @@
     "axios": "^0.23.0",
     "bootstrap": "^4.6.0",
     "bootstrap-vue": "^2.21.2",
-    "chart.js": "^2.9.4",
+    "chart.js": "2.9.4",
```
Comment (AndreiNavoichyk) [inline frontend/package.json:42]: Why did you remove hat?

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue [hunk 11]
```diff
@@ -0,0 +1,585 @@
+<template>
```
Comment (AndreiNavoichyk) [inline frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: Rename to `CumulativeSignedExtrinsicsChart`

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue [hunk 12]
```diff
@@ -0,0 +1,585 @@
...
+  <div>
+    <div v-if="loading" class="text-center py-4">
+      <Loading />
+    </div>
+    <div v-else>
+      <b-button-group size="sm" class="mt-5">
```
Comment (AndreiNavoichyk) [inline frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: It should be a separate component so you can reuse it.

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+        },
+        animation: {
+          duration: 300,
+        },
+      },
+    }
+  },
+  methods: {
+    month() {
+      this.chartOptions = {
```
Comment (AndreiNavoichyk) [inline frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: `chartOptions` should be reused.

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue [hunk 14]
```diff
@@ -0,0 +1,585 @@
...
+      },
+    },
+  },
+}
+</script>
+<style scoped>
+.btn {
+  background: none;
+  border: solid 1px black;
+}
```
Comment (AndreiNavoichyk) [inline frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: Add a newline as a separator between styles.

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue (hunk 11 above)
Comment (AndreiNavoichyk) [reply frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: How does it relate to `validator`? Move it to the right folder.

File: backend/docker/sql/views/signed_extrinsics_per_day.sql (hunk 1 above)
Comment (karthikeyan-cere) [reply backend/docker/sql/views/signed_extrinsics_per_day.sql]: added

File: backend/docker/sql/views/signed_extrinsics_per_month.sql (hunk 4 above)
Comment (karthikeyan-cere) [reply backend/docker/sql/views/signed_extrinsics_per_month.sql]: added

File: frontend/package.json (hunk 10 above)
Comment (karthikeyan-cere) [reply frontend/package.json:42]: In order to install 2.9.4, the updated version won't work with our configuration

File: frontend/locales/en.json (hunk 5 above)
Comment (karthikeyan-cere) [reply frontend/locales/en.json]: updated

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue (hunk 13 above)
Comment (karthikeyan-cere) [reply frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: updated

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue (hunk 14 above)
Comment (karthikeyan-cere) [reply frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: added

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue (hunk 11 above)
Comment (karthikeyan-cere) [reply frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: Moved to charts folder inside components

File: backend/docker/sql/views/signed_extrinsics_per_day.sql (hunk 2 above)
Comment (karthikeyan-cere) [reply backend/docker/sql/views/signed_extrinsics_per_day.sql]: updated

File: backend/docker/sql/views/signed_extrinsics_per_day.sql (hunk 1 above)
Comment (AndreiNavoichyk) [reply backend/docker/sql/views/signed_extrinsics_per_day.sql]: Add to the view name as well. Not only file.

File: backend/docker/sql/views/signed_extrinsics_per_month.sql (hunk 4 above)
Comment (AndreiNavoichyk) [reply backend/docker/sql/views/signed_extrinsics_per_month.sql]: Add to the view name as well. Not only file.

File: backend/docker/sql/views/signed_extrinsics_per_day.sql (hunk 3 above)
Comment (karthikeyan-cere) [reply backend/docker/sql/views/signed_extrinsics_per_day.sql:8]: Can't use variables in psql.

File: frontend/locales/en.json (hunk 7 above)
Comment (karthikeyan-cere) [reply frontend/locales/en.json]: updated

File: backend/docker/sql/views/signed_extrinsics_per_month.sql (hunk 4 above)
Comment (karthikeyan-cere) [reply backend/docker/sql/views/signed_extrinsics_per_month.sql]: added

File: frontend/components/validator/charts/CumulativeExtrinsicsChart.vue (hunk 12 above)
Comment (karthikeyan-cere) [reply frontend/components/validator/charts/CumulativeExtrinsicsChart.vue]: Done

File: frontend/components/charts/CumulativeSignedExtrinsicsChart.vue [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+export default {
+  components: {
+    ReactiveLineChart,
+    Loading,
+    ChartFilter,
+  },
+  mixins: [signedExtrinsic],
+  data() {
+    return {
+      filterButtons: [
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/CumulativeSignedExtrinsicsChart.vue]: Let's move these buttons as default ones to the filter component.

File: frontend/components/charts/SignedExtrinsicsChart.vue [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+export default {
+  components: {
+    BarChart,
+    Loading,
+    ChartFilter,
+  },
+  mixins: [signedExtrinsic],
+  data() {
+    return {
+      filterButtons: [
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/SignedExtrinsicsChart.vue]: You should reuse this code between charts. Think about factory methods pattern

File: frontend/components/charts/CumulativeSignedExtrinsicsChart.vue [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+  },
+  methods: {
+    async month() {
+      this.activeButton = '30D'
+      const queryLimit = 30
+      const { count, label } = await this.extrinsicsDayCount(queryLimit)
+      const cumulativeCount = await this.cumulativeValue(count)
+      this.chartData = {
+        labels: label,
+        datasets: [
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/CumulativeSignedExtrinsicsChart.vue]: This code should be reused.

**Review by AndreiNavoichyk — CHANGES_REQUESTED** [review]

File: frontend/components/charts/SignedExtrinsicsChart.vue [hunk 18]
```diff
@@ -0,0 +1,268 @@
+<template>
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/SignedExtrinsicsChart.vue]: Rename to `SignedExtrinsicsBarChart`

File: frontend/components/charts/CumulativeSignedExtrinsicsChart.vue [hunk 19]
```diff
@@ -0,0 +1,272 @@
+<template>
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/CumulativeSignedExtrinsicsChart.vue]: Rename `SignedExtrinsicsCumulativeChart`

File: frontend/components/charts/SignedExtrinsicsChart.vue (hunk 18 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/SignedExtrinsicsChart.vue]: done

File: frontend/components/charts/CumulativeSignedExtrinsicsChart.vue (hunk 19 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/CumulativeSignedExtrinsicsChart.vue]: done

File: frontend/components/charts/CumulativeSignedExtrinsicsChart.vue (hunk 17 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/CumulativeSignedExtrinsicsChart.vue]: done

File: frontend/components/charts/SignedExtrinsicsChart.vue (hunk 16 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/SignedExtrinsicsChart.vue]: done

File: frontend/components/charts/CumulativeSignedExtrinsicsChart.vue (hunk 15 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/CumulativeSignedExtrinsicsChart.vue]: Moved to mixing

File: frontend/components/charts/SignedExtrinsicsBarChart.vue [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+      this.activeButton = 'Max'
+      const queryLimit = 12
+      const { count, label } = await this.extrinsicsMonthCount(queryLimit)
+      const chartData = this.getChartData(label, count)
+      this.chartData = chartData
+      this.loading = false
+    },
+  },
+}
+</script>
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/SignedExtrinsicsBarChart.vue:85]: Styles should be reused.

File: frontend/mixins/signedExtrinsic.js [hunk 21]
```diff
@@ -0,0 +1,166 @@
+import { gql } from 'graphql-tag'
```
Comment (AndreiNavoichyk) [inline frontend/mixins/signedExtrinsic.js]: Rename this file so we can collect all charts related logic here.

File: frontend/mixins/signedExtrinsic.js [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+export default {
+  methods: {
+    async extrinsicsDayCount(queryLimit) {
```
Comment (AndreiNavoichyk) [inline frontend/mixins/signedExtrinsic.js]: Rename all methods so it is clear that it is about signed trxs

File: frontend/mixins/signedExtrinsic.js [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+              },
+            },
+          ],
+        },
+        animation: {
+          duration: 300,
+        },
+      }
+    },
+    getFilterButtons() {
```
Comment (AndreiNavoichyk) [inline frontend/mixins/signedExtrinsic.js]: It should be in the component as a default value.

File: frontend/components/charts/SignedExtrinsicsBarChart.vue (hunk 20 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/SignedExtrinsicsBarChart.vue:85]: done

File: frontend/mixins/signedExtrinsic.js (hunk 22 above)
Comment (karthikeyan-cere) [reply frontend/mixins/signedExtrinsic.js]: Done

File: frontend/mixins/signedExtrinsic.js (hunk 23 above)
Comment (karthikeyan-cere) [reply frontend/mixins/signedExtrinsic.js]: Can't able to move to chart filter component, as we can't bind methods in parent component,

File: CHANGELOG.md [hunk 24]
```diff
@@ -2,6 +2,7 @@
 ## vNext
 
 - Removed auto-payout and participation in governance filters
+- Added signed extrinsic cumulative and bar charts
```
Comment (AndreiNavoichyk) [inline CHANGELOG.md]: Too much new lines.

File: CHANGELOG.md [hunk 25]
```diff
@@ -2,6 +2,7 @@
...
```
Comment (AndreiNavoichyk) [inline CHANGELOG.md]: Let's split into 2:
1. Implemented charts framework for cumulative line and bar types
2. Implemented `Signed Extrinsics` charts.

File: frontend/mixins/chartMixin.js [hunk 26]
```diff
@@ -0,0 +1,166 @@
+import { gql } from 'graphql-tag'
```
Comment (AndreiNavoichyk) [inline frontend/mixins/chartMixin.js]: Rename to `charts*`

File: frontend/mixins/chartMixin.js [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+    },
+    cumulativeValue(value) {
+      return value.map(
+        (
+          (sum) => (value) =>
+            (sum += value)
+        )(0)
+      )
+    },
+    getChartData(label, data) {
```
Comment (AndreiNavoichyk) [inline frontend/mixins/chartMixin.js]: Take a look to the body. You have specific there. Rename the method.

File: frontend/mixins/chartMixin.js [hunk 28]
```diff
@@ -0,0 +1,166 @@
...
+              },
+            },
+          ],
+        },
+        animation: {
+          duration: 300,
+        },
+      }
+    },
+    getFilterButtons() {
```
Comment (AndreiNavoichyk) [inline frontend/mixins/chartMixin.js]: Rename to `getDefault*`

File: frontend/mixins/signedExtrinsic.js (hunk 21 above)
Comment (karthikeyan-cere) [reply frontend/mixins/signedExtrinsic.js]: done

File: CHANGELOG.md (hunk 25 above)
Comment (karthikeyan-cere) [reply CHANGELOG.md]: done

File: CHANGELOG.md (hunk 24 above)
Comment (karthikeyan-cere) [reply CHANGELOG.md]: removed additional spaces

File: frontend/mixins/chartMixin.js (hunk 26 above)
Comment (karthikeyan-cere) [reply frontend/mixins/chartMixin.js]: Renamed

File: frontend/mixins/chartMixin.js (hunk 28 above)
Comment (karthikeyan-cere) [reply frontend/mixins/chartMixin.js]: Renamed

File: frontend/mixins/chartMixin.js (hunk 27 above)
Comment (karthikeyan-cere) [reply frontend/mixins/chartMixin.js]: Renamed as `getExtrinsicsChartData`

File: frontend/components/charts/SignedExtrinsicsBarChart.vue [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+        'Extrinsics per Period'
+      ),
+    }
+  },
+  mounted() {
+    this.month()
+  },
+  methods: {
+    async month() {
+      this.activeButton = '30D'
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/SignedExtrinsicsBarChart.vue]: Move this value to constant and reuse. The same for all buttons.

File: frontend/components/charts/SignedExtrinsicsBarChart.vue [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+      const queryLimit = 30
```
Comment (AndreiNavoichyk) [inline frontend/components/charts/SignedExtrinsicsBarChart.vue]: Move this value to constant and reuse. The same for all buttons.

File: frontend/components/charts/SignedExtrinsicsBarChart.vue (hunk 29 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/SignedExtrinsicsBarChart.vue]: Done

File: frontend/components/charts/SignedExtrinsicsBarChart.vue (hunk 30 above)
Comment (karthikeyan-cere) [reply frontend/components/charts/SignedExtrinsicsBarChart.vue]: Done

**Review by AndreiNavoichyk — APPROVED** [review]
